// @makerdao/dai/dist/src/eth/tokens/PethToken.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports["default"] = void 0;

var _regenerator = _interopRequireDefault(require("@babel/runtime/regenerator"));

var _asyncToGenerator2 = _interopRequireDefault(require("@babel/runtime/helpers/asyncToGenerator"));

var _classCallCheck2 = _interopRequireDefault(require("@babel/runtime/helpers/classCallCheck"));

var _createClass2 = _interopRequireDefault(require("@babel/runtime/helpers/createClass"));

var _inherits2 = _interopRequireDefault(require("@babel/runtime/helpers/inherits"));

var _possibleConstructorReturn2 = _interopRequireDefault(require("@babel/runtime/helpers/possibleConstructorReturn"));

var _getPrototypeOf2 = _interopRequireDefault(require("@babel/runtime/helpers/getPrototypeOf"));

var _Erc20Token2 = _interopRequireDefault(require("./Erc20Token"));

var _Currency = require("../Currency");

function _createSuper(Derived) { var hasNativeReflectConstruct = _isNativeReflectConstruct(); return function () { var Super = (0, _getPrototypeOf2["default"])(Derived), result; if (hasNativeReflectConstruct) { var NewTarget = (0, _getPrototypeOf2["default"])(this).constructor; result = Reflect.construct(Super, arguments, NewTarget); } else { result = Super.apply(this, arguments); } return (0, _possibleConstructorReturn2["default"])(this, result); }; }

function _isNativeReflectConstruct() { if (typeof Reflect === "undefined" || !Reflect.construct) return false; if (Reflect.construct.sham) return false; if (typeof Proxy === "function") return true; try { Date.prototype.toString.call(Reflect.construct(Date, [], function () {})); return true; } catch (e) { return false; } }

var PethToken = /*#__PURE__*/function (_Erc20Token) {
  (0, _inherits2["default"])(PethToken, _Erc20Token);

  var _super = _createSuper(PethToken);

  function PethToken(contract, web3Service, tub) {
    var _this;

    (0, _classCallCheck2["default"])(this, PethToken);
    _this = _super.call(this, contract, web3Service, 18, 'PETH');
    _this._tub = tub;
    return _this;
  }

  (0, _createClass2["default"])(PethToken, [{
    key: "join",
    value: function join(amount) {
      var _ref = arguments.length > 1 && arguments[1] !== undefined ? arguments[1] : {},
          _ref$unit = _ref.unit,
          unit = _ref$unit === void 0 ? _Currency.WETH : _ref$unit,
          promise = _ref.promise;

      var value = this._valueForContract(amount, unit);

      return this._tub.join(value, {
        promise: promise
      });
    }
  }, {
    key: "exit",
    value: function exit(amount) {
      var _ref2 = arguments.length > 1 && arguments[1] !== undefined ? arguments[1] : {},
          _ref2$unit = _ref2.unit,
          unit = _ref2$unit === void 0 ? _Currency.PETH : _ref2$unit,
          promise = _ref2.promise;

      var value = this._valueForContract(amount, unit);

      return this._tub.exit(value, {
        promise: promise
      });
    }
  }, {
    key: "wrapperRatio",
    value: function () {
      var _wrapperRatio = (0, _asyncToGenerator2["default"])( /*#__PURE__*/_regenerator["default"].mark(function _callee() {
        return _regenerator["default"].wrap(function _callee$(_context) {
          while (1) {
            switch (_context.prev = _context.next) {
              case 0:
                _context.t0 = _Currency.WETH;
                _context.next = 3;
                return this._tub.per();

              case 3:
                _context.t1 = _context.sent;
                return _context.abrupt("return", _context.t0.ray.call(_context.t0, _context.t1));

              case 5:
              case "end":
                return _context.stop();
            }
          }
        }, _callee, this);
      }));

      function wrapperRatio() {
        return _wrapperRatio.apply(this, arguments);
      }

      return wrapperRatio;
    }()
  }, {
    key: "joinPrice",
    value: function () {
      var _joinPrice = (0, _asyncToGenerator2["default"])( /*#__PURE__*/_regenerator["default"].mark(function _callee2(amount) {
        var unit,
            value,
            _args2 = arguments;
        return _regenerator["default"].wrap(function _callee2$(_context2) {
          while (1) {
            switch (_context2.prev = _context2.next) {
              case 0:
                unit = _args2.length > 1 && _args2[1] !== undefined ? _args2[1] : _Currency.WETH;
                value = this._valueForContract(amount, unit);
                _context2.t0 = _Currency.WETH;
                _context2.next = 5;
                return this._tub.ask(value);

              case 5:
                _context2.t1 = _context2.sent;
                return _context2.abrupt("return", _context2.t0.wei.call(_context2.t0, _context2.t1));

              case 7:
              case "end":
                return _context2.stop();
            }
          }
        }, _callee2, this);
      }));

      function joinPrice(_x) {
        return _joinPrice.apply(this, arguments);
      }

      return joinPrice;
    }()
  }, {
    key: "exitPrice",
    value: function () {
      var _exitPrice = (0, _asyncToGenerator2["default"])( /*#__PURE__*/_regenerator["default"].mark(function _callee3(amount) {
        var unit,
            value,
            _args3 = arguments;
        return _regenerator["default"].wrap(function _callee3$(_context3) {
          while (1) {
            switch (_context3.prev = _context3.next) {
              case 0:
                unit = _args3.length > 1 && _args3[1] !== undefined ? _args3[1] : _Currency.WETH;
                value = this._valueForContract(amount, unit);
                _context3.t0 = _Currency.WETH;
                _context3.next = 5;
                return this._tub.bid(value);

              case 5:
                _context3.t1 = _context3.sent;
                return _context3.abrupt("return", _context3.t0.wei.call(_context3.t0, _context3.t1));

              case 7:
              case "end":
                return _context3.stop();
            }
          }
        }, _callee3, this);
      }));

      function exitPrice(_x2) {
        return _exitPrice.apply(this, arguments);
      }

      return exitPrice;
    }()
  }]);
  return PethToken;
}(_Erc20Token2["default"]);

exports["default"] = PethToken;
